main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -0,0 +1,188 @@
+import ca.uwaterloo.flix.util.collection.CofiniteSet
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+import java.util.concurrent.ConcurrentLinkedQueue
+
+/**
+  * Rewrites the body of each def using, applying the following transformations:
+  *   - Copy Propagation:
+  * {{{
+  *     let x = 1;
+  *     f(x)
+  * }}}
+  *     becomes
+  * {{{
+  *     let x = 1;
+  *    f(1)
+  * }}}
+  *   - Dead Code Elimination
+  * {{{
+  *     let x = 1;
+  *     f(1)
+  * }}}
+  *     becomes
+  * {{{
+  *     f(1)
+  * }}}
+  *   - Inline Expansion
+  * {{{
+  *     f(1)
+  * }}}
+  *     becomes (where the definition of `f` is `x + 2`)
+  * {{{
+  *     (x -> x + 2)(1)
+  * }}}
+  *   - Beta Reduction
+  * {{{
+  *     (x -> x + 2)(1)
+  * }}}
+  *     becomes
+  * {{{
+  *     let x = 1;
+  *     x + 2
+  * }}}
+  */
+object Inliner {
+
+  /** Performs inlining on the given AST `root`. */
+  def run(root: OccurrenceAst.Root)(implicit flix: Flix): OccurrenceAst.Root = {
+    val defs = ParOps.parMapValues(root.defs)(visitDef(_)(root, flix))
+    val newRoot = OccurrenceAst.Root(defs, root.enums, root.structs, root.effects, root.mainEntryPoint, root.entryPoints, root.sources)
+    newRoot
+  }
+
+  /** Performs inlining on the body of `def0`. */
+  private def visitDef(def0: OccurrenceAst.Def)(implicit root: OccurrenceAst.Root, flix: Flix): OccurrenceAst.Def = def0 match {
+    case OccurrenceAst.Def(sym, fparams, spec, exp, ctx, loc) =>
+      val e = visitExp(exp, Context.Empty)(sym, root, flix)
+      OccurrenceAst.Def(sym, fparams, spec, e, ctx, loc)
+  }
+
+  /** Performs inlining operations on the expression `exp0` from [[Expr]]. */
+  private def visitExp(exp0: Expr, ctx0: Context)(implicit sym0: Symbol.DefnSym, root: OccurrenceAst.Root, flix: Flix): Expr = exp0
+
+  /** Returns `true` if `def0` should be inlined. */
+  private def shouldInline(defCtx: DefContext, ctx: Context): Boolean = {
+    val mayInline = !defCtx.isSelfRecursive && !ctx.currentlyInlining
+    val shouldInline = defCtx.isDirectCall
+    mayInline && shouldInline
+  }
+
+  /** Returns `true` if `eff0` is pure. */
+  private def isPure(eff0: Type): Boolean = {
+    eff0 == Type.Pure
+  }
+
+  /** Checks if `occur` is [[Dead]]. */
+  private def isDead(occur: OccurrenceAst.Occur): Boolean = occur match {
+    case Occur.Dead => true
+    case Occur.Once => false
+    case Occur.OnceInLambda => false
+    case Occur.OnceInLocalDef => false
+    case Occur.ManyBranch => false
+    case Occur.Many => false
+  }
+
+  /** Checks if `occur` is [[Once]] and `eff` is pure. */
+  private def isUsedOnceAndPure(occur: OccurrenceAst.Occur, eff0: Type): Boolean = occur match {
+    case Once => isPure(eff0)
+    case _ => false
+  }
+
+  /** Checks if `exp0` is trivial and `eff` is pure */
+  private def isTrivialAndPure(exp0: Expr, eff0: Type): Boolean = {
+    isPure(eff0) && isTrivialExp(exp0)
```

Comment:
name is "isTrivialAndPure" but we check in opposite order. IT does not matter but is still strange.

Reply:
I think we should just remove all of these `xAndY` predicates and just use `&&` at call sites.

Reply:
Yes, agreed.